File: ios/Oriveo/Oriveo/Core/Attachments/AttachmentImportLimiter.swift
import Foundation

nonisolated struct AttachmentImportLimiter: Sendable {
    nonisolated struct Result: Sendable {
        let accepted: [Attachment]
        let rejectedCount: Int
    }

    nonisolated static func limit(
        existing: [Attachment],
        incoming: [Attachment],
        maxAttachments: Int
    ) -> Result {
        guard maxAttachments > 0 else {
            return Result(accepted: [], rejectedCount: incoming.count)
        }

        let availableSlots = max(maxAttachments - existing.count, 0)
        guard availableSlots > 0 else {
            return Result(accepted: [], rejectedCount: incoming.count)
        }

        if incoming.count <= availableSlots {
            return Result(accepted: incoming, rejectedCount: 0)
        }

        return Result(
            accepted: Array(incoming.prefix(availableSlots)),
            rejectedCount: incoming.count - availableSlots
        )
    }
}

File: android/app/src/main/java/ai/oriveo/community/core/attachments/AttachmentImportLimiter.kt
package ai.oriveo.community.core.attachments

import ai.oriveo.community.core.model.Attachment

object AttachmentImportLimiter {

    data class Result(
        val accepted: List<Attachment>,
        val rejectedCount: Int,
    )

    fun limit(
        existing: List<Attachment>,
        incoming: List<Attachment>,
        maxAttachments: Int,
    ): Result {
        if (maxAttachments <= 0) {
            return Result(accepted = emptyList(), rejectedCount = incoming.size)
        }

        val availableSlots = (maxAttachments - existing.size).coerceAtLeast(0)
        if (availableSlots == 0) {
            return Result(accepted = emptyList(), rejectedCount = incoming.size)
        }
        if (incoming.size <= availableSlots) {
            return Result(accepted = incoming, rejectedCount = 0)
        }
        return Result(
            accepted = incoming.take(availableSlots),
            rejectedCount = incoming.size - availableSlots,
        )
    }
}
